File: codeforces/1023/D/main.py
import math
t = input().split()
n, q = int(t[0]), int(t[1])
a = [int(s) for s in input().split()]
rawA = a.copy()
hasZero = False
allAreZero = True
lastNonZero = 1
firstZero = 0
for i in range(n):
    if a[i] != 0:
        if allAreZero:
            for j in range(i):
                a[j] = a[i]
        lastNonZero = a[i]
        allAreZero = False
    else:
        hasZero = True
        firstZero = i
        if not allAreZero:
            a[i] = lastNonZero
if allAreZero:
    a = [q] * n
    print("YES")
    print(" ".join([str(c) for c in a]))
    exit()

rmq = []
rmq.append(a)
for k in range(1, int(math.log(n, 2) + 1)):
    t = rmq[k-1]
    x = []
    for i in range(n):
        r = min(i + pow(2, k - 1), n - 1)
        x.append(min(t[i], t[r]))
    rmq.append(x)

def query(ll, rr):
    if ll > rr:
        return q + 1
    k = int(math.log(rr - ll + 1, 2))
    return min(rmq[k][ll], rmq[k][rr - pow(2, k) + 1])

l = {}
r = {}
for i in range(n):
    if not a[i] in l:
        l[a[i]] = i
    r[a[i]] = i

for x in range(1, q + 1):
    if x in l:
        if query(l[x]+1, r[x]-1) < x:
            print("NO")
            exit()

if not q in l:
    if not hasZero:
        print("NO")
        exit()
    else:
        a[firstZero] = q

print("YES")
for c in a:
    print(c, end = ' ')

File: codeforces/1023/D/main.cpp
#include <bits/stdc++.h>
using namespace std;
int n, q;
vector<int> a;
bool hasZero = false;
bool allAreZero = true;
int lastNonZero = 1;
int firstZero = 0;

int rmq[32][1000000];
int l[1000000];
int r[1000000];
int query(int l, int r)
{
    if (l > r)
        return q + 1;
    int k = log2(r - l + 1);
    return min(rmq[k][l], rmq[k][r - (1 << k) + 1]);
}

int main()
{
    scanf("%d%d", &n, &q);
    for (int i = 0; i < n; i++)
    {
        int t;
        scanf("%d", &t);
        a.push_back(t);
    }
    for (int i = 0; i < n; i++)
    {
        if (a[i] != 0)
        {
            if (allAreZero)
            {
                for (int j = 0; j < i; j++)
                {
                    a[j] = a[i];
                }
            }
            allAreZero = false;
            lastNonZero = a[i];
        }
        else
        {
            hasZero = true;
            firstZero = i;
            if (!allAreZero)
            {
                a[i] = lastNonZero;
            }
        }
    }
    if (allAreZero)
    {
        puts("YES");
        for (int i = 0; i < n; i++)
            printf("%d ", q);
        return 0;
    }
    for (int i = 0; i < n; i++)
        rmq[0][i] = a[i];
    for (int k = 1; k < log2(n) + 1; k++)
    {
        for (int i = 0; i < n; i++)
        {
            int rIndex = min(i + (1 << (k - 1)), n - 1);
            rmq[k][i] = min(rmq[k - 1][i], rmq[k - 1][rIndex]);
        }
    }
    memset(l, -1, sizeof(l));
    memset(r, -1, sizeof(r));
    for (int i = 0; i < n; i++)
    {
        int t = a[i];
        if (l[t] == -1)
            l[t] = i;
        r[t] = i;
    }
    for (int x = 1; x <= q; x++)
    {
        if (l[x] != -1)
        {
            if (query(l[x] + 1, r[x] - 1) < x)
            {
                puts("NO");
                return 0;
            }
        }
    }
    if (l[q] == -1)
    {
        if (!hasZero)
        {
            puts("NO");
            return 0;
        }
        else
        {
            a[firstZero] = q;
        }
    }
    puts("YES");
    for(int t: a) printf("%d ", t);
    return 0;
}